pilot/pkg/model/service.go

```diff
@@ -477,6 +490,23 @@ func ParseServiceKey(s string) (hostname string, ports PortList, labels LabelsCo
 	return
 }
 
+// BuildSubsetKey generates a unique string referencing service instances for a given service name, a subset and a port.
+// The proxy queries Pilot with this key to obtain the list of instances in a subset.
+func BuildSubsetKey(direction TrafficDirection, subsetName, hostname string, port *Port) string {
```

Comment:
I think we should discuss this a bit more - as I mentioned in another PR, we should have a clear reason for making things more complex than they need to be.

Why not use DNS:port ( and :port for inbound) ? 

AFAIK subsets are DNS names, and may have associated DNS entries in some cases.

Reply:
the reason for having subset name for inbound is so that in the dashboards, we can see the amount of traffic that goes to a service, and at the same time, we can also correlate with the amount of traffic entering the service (as recorded by the inbound....). Hence the need to have inbound and outbound.

With regard to ':', its not an allowed character in Envoy, for cluster names.

Reply:
I am not opposed to creating the cluster name as a extended FQDN name. It increases the parsing time though. Also, if there is no subset, how do we distinguish the case where a subset name is missing (used for default clusters)? Having a simple dumb a|b|c|d string allows us to easily parse the components in one string split.

Also, it lends itself well to dashboards, that use regexes.. cc @douglas-reid

Reply:
in general, my interests are only in having a straightforward way to trim down cluster names into something that is user-friendly in the places that we will use Envoy metrics directly. Currently, there aren't a ton of those places, so this should be a minor concern.

As you say, '.' is problematic. Anything else, '|' or otherwise seems fine by me.

Reply:
The benefit of using a DNS format outweigh the parsing problem. BTW, why is "." harder to parse than "."?

Reply:
@andraxylia: Requires slightly more logic.
`outbound|http|subsetx|uk.abcdef.com`
vs
`outbound.http.subsetx.uk.abcdef.com`
To get the components from the 1st one split on `|`
To get the components from the 2nd one split on `.` and the hostname is `parts[3:]`
Easy to do in code, harder to do with metrics tools.

Reply:
also, the metrics as exported (in statsd format) by envoy use '.' to separate other parts:

See: https://www.envoyproxy.io/docs/envoy/v1.6.0/configuration/cluster_manager/cluster_stats

> Every cluster has a statistics tree rooted at cluster.\<name\>. 

This gets worse when you consider:

> If HTTP is used, dynamic HTTP response code statistics are also available. These are emitted by various internal systems as well as some filters such as the router filter and rate limit filter. They are rooted at cluster.\<name\>. and contain the following statistics: